### DISCUSSION TITLE
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

### DISCUSSION THREAD

**Discussion by zhushenhang** [react-hook-form/react-hook-form#3744] [Q&A] [answered]
hello.
the complete warning message is as follows:
`index.js:1406 Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

Check the render method of `Controller`.
    in BorderedTextField (created by Controller)
    in Controller (at LoginForm.js:13)
    in form (at LoginForm.js:12)
    in LoginForm (at Login.js:53)
    in div (created by ForwardRef(Paper))
    in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
    in WithStyles(ForwardRef(Paper)) (at Login.js:48)
    in div (at Login.js:47)
    in Login (created by inject-with-alertStore(Login))
    in inject-with-alertStore(Login) (created by InnerComponent)
    in InnerComponent
    in ErrorBoundary
    in Unknown (created by WithStyles(Component))
    in WithStyles(Component) (created by Context.Consumer)
    in Route (at App.js:18)
    in Switch (at App.js:15)
    in Router (created by BrowserRouter)
    in BrowserRouter (at App.js:13)
    in MobXProvider (at App.js:12)
    in App (at src/index.js:6)`

my form code.
`<form onSubmit={handleSubmit(onSubmit)} className={formClassName}>
      <Controller
        as={BorderedTextField}
        name="userName"
        label="userName"
        defaultValue=""
        autoFocus
        margin="normal"
        fullWidth
        required
        control={control}
        rules={{ required: true, pattern: /^[\w]{1,16}$/ }}
        helperText={errors.userName && 'Please try again'}
        error={errors.userName ? true : false}
      />
      <Controller
        as={BorderedTextField}
        name="password"
        label="password"
        type="password"
        defaultValue=""
        margin="normal"
        fullWidth
        required
        control={control}
        rules={{ required: true, pattern: /^[\w]{4,16}$/ }}
        helperText={errors.password && 'Please try again'}
        error={errors.password ? true : false}
      />
      <Button
        fullWidth
        variant="contained"
        color="primary"
        className={buttonClassName}
        type="submit"
      >
        <Typography>submit</Typography>
      </Button>
    </form>` 

i do not use ref and BorderedTextField base on Material-ui.
version ^6.11.0.
can you tell me the reason for the warning
thanks.

**Comment by bluebill1049** [comment, marked as answer]
looks like your external component doesn't support `ref` prop, you can use `render` prop to omit the `ref`

`render({ ref, ...rest })` at Controller

**Reply by zhushenhang** [reply]
You're awesome.
Thank you very much

**Reply by stephent** [reply]
@bluebill1049 is there an example of how to implement support for the `ref` prop in a custom component intended for use with `Controller`? I'm not 100% clear on how it should be wired up, or what the consequences of omitting it would be.

E.g. in something like this:

```js
import React from 'react'

export default function CustomComponent( { onChange, onBlur, value } ) {
  
  return (
    <div className="btn-group">
      <button type="button" onClick={() => onChange(-1)}>
	<img src={...} />
      </button>
      <button type="button" onClick={() => onChange(0)}>
	<img src={...} />
      </button>
      <button type="button" onClick={() => onChange(1)}>
	<img src={...} />
      </button>
    </div>
  );
}
```

Thanks!

**Reply by bluebill1049** [reply]
Have a look at this section: https://react-hook-form.com/advanced-usage#ExtendController It may helps.

**Comment by Syknapse** [comment]
If anyone stumbles upon this after v7 and is wondering how to apply this with the new controller syntax of `({ field, fieldState })` this is how I applied it
```
 render={({ field: { ref, ...rest }, fieldState }) => (
    <PhoneInput {...rest} {...fieldState} error={errors.phone} />
  )}
```

**Reply by wlsp** [reply]
Thanks a lot mate! Just made my day!

**Reply by jayYoung888** [reply]
Thanks a ton for your work!
